src/clr-angular/data/datagrid/render/row-renderer.ts

```diff
@@ -3,29 +3,54 @@
  * This software is released under MIT license.
  * The full license information can be found in LICENSE in the root directory of this project.
  */
-import { AfterViewInit, ContentChildren, Directive, QueryList } from '@angular/core';
+import { AfterContentInit, ContentChildren, Directive, OnDestroy, QueryList } from '@angular/core';
 
 import { DatagridCellRenderer } from './cell-renderer';
 import { ColumnsService } from '../providers/columns.service';
+import { Subscription } from 'rxjs/index';
 
 @Directive({ selector: 'clr-dg-row, clr-dg-row-detail' })
-export class DatagridRowRenderer implements AfterViewInit {
+export class DatagridRowRenderer implements AfterContentInit, OnDestroy {
   @ContentChildren(DatagridCellRenderer) private cells: QueryList<DatagridCellRenderer>;
 
   constructor(private columnsService: ColumnsService) {}
 
-  ngAfterViewInit() {
-    this.setColumnStates();
-    this.cells.changes.subscribe(() => {
-      this.setColumnStates();
-    });
+  ngAfterContentInit() {
+    this.setColumnStates(); // case #3 and #4
+    this.subscriptions.push(
+      this.cells.changes.subscribe(() => {
+        this.setColumnStates(); // case #2
+        // Note on case #2: In the case of dynamic columns, when one column (header/cell together) gets deleted,
+        // this.cells.changes fires before this.columnsService.columns gets updated.
+        // So that means there will be more an extra dead service from the deleted column would hanging around.
+        // Hence, we should quit earlier here, but this method will be called again right
+        // after this.columnsService.columns gets updated properly, in which the method will run completely.
+      })
+    );
+  }
+
+  private subscriptions: Subscription[] = [];
+
+  ngOnDestroy() {
+    this.subscriptions.forEach(sub => sub.unsubscribe());
   }
 
   public setColumnStates() {
-    this.cells.forEach((cell, index) => {
-      if (this.columnsService.columns[index]) {
-        cell.columnState = this.columnsService.columns[index];
-      }
-    });
+    // This method runs in four cases:
+    // 1. When the initial rows appear on the first page.
+    //    In this case, the method will be called in DatagridMainRenderer.
+    // 2. When columns (corresponding header/cells) get added and deleted.
+    //    In this case, the method will be called in DatagridMainRenderer. (Read the note on this case above).
+    // 3. When rows load asynchronously.
+    //    In this case, the method will be called in this class.
+    // 4. When rows load after switching pages.
+    //    In this case, the method will be called in this class (Basically, same as the case 3).
+    if (this.cells.length === this.columnsService.columns.length) {
```

Comment:
Don't we have a potential issue here, where we hide a column and service would have n+1 columns in it?

Reply:
It is right here, this guards that only when the headers have stabilized do we run this.